From 0ef134636e09340278c8cac6cf402dbfdce3ccfc Mon Sep 17 00:00:00 2001
From: Corinna Vinschen <vinschen@redhat.com>
Date: Sat, 12 Feb 2022 17:59:04 +0100
Subject: [PATCH 5/6] if WinHello device is present, use it exclusively

Disable usage of direct USB device access for admin users on systems
supporting WinHello.  The actual device is the same anyway, just handled
via the WinHello abstraction layer.  This also avoids calling sk_try if
WinHello is present.

Signed-off-by: Corinna Vinschen <vinschen@redhat.com>
---
 sk-usbhid.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/sk-usbhid.c b/sk-usbhid.c
index e906eb3d4999..98b69ad4e362 100644
--- a/sk-usbhid.c
+++ b/sk-usbhid.c
@@ -504,6 +504,13 @@ sk_select_by_cred(const fido_dev_info_t *devlist, size_t ndevs,
 		skv[0] = NULL;
 		goto out;
 	}
+	for (i = 0; i < skvcnt; i++) {
+		if (fido_dev_is_winhello (skv[i]->dev)) {
+			sk = skv[i];
+			skv[i] = NULL;
+			goto out;
+		}
+	}
 	sk = NULL;
 	for (i = 0; i < skvcnt; i++) {
 		if (sk_try(skv[i], application, key_handle,
@@ -540,6 +547,13 @@ sk_select_by_touch(const fido_dev_info_t *devlist, size_t ndevs)
 		}
 		goto out;
 	}
+	for (idx = 0; idx < skvcnt; idx++) {
+		if (fido_dev_is_winhello (skv[idx]->dev)) {
+			sk = skv[idx];
+			skv[idx] = NULL;
+			goto out;
+		}
+	}
 #ifndef HAVE_FIDO_DEV_GET_TOUCH_STATUS
 	skdebug(__func__, "libfido2 version does not support a feature needed for multiple tokens. Please upgrade to >=1.5.0");
 	goto out;
-- 
2.35.1

